core/src/executor/evaluate/evaluated.rs

```diff
@@ -210,22 +313,469 @@ impl<'a> Evaluated<'a> {
+        if matched_vec[0].0 != 0 && matched_vec[matched_vec.len() - 1].0 != sliced_expr.len() - 1 {
+            return Ok(Evaluated::StrSlice { source, range });
+        }
+
+        let pivot = matched_vec
+            .iter()
+            .rev()
+            .enumerate()
+            .skip_while(|(vec_idx, (slice_idx, _))| *vec_idx == sliced_expr.len() - slice_idx - 1)
+            .map(|(vec_idx, (_, _))| vec_idx)
+            .next();
+
+        let end = match pivot {
+            Some(idx) => match idx {
+                0 => range.end,
+                _ => matched_vec[matched_vec.len() - idx].0,
+            },
+            _ => matched_vec[0].0,
+        };
+
+        Ok(Evaluated::StrSlice {
+            source: expr_str.to_owned(),
+            range: range.start..end,
+        })
+    }
+
+    pub fn substr(
+        self,
+        name: String,
+        start: Evaluated<'a>,
+        count: Option<Evaluated<'a>>,
+    ) -> Result<Evaluated<'a>> {
+        let (source, range) = match self {
+            Evaluated::Literal(literal) => {
+                let value = Value::try_from(literal)?;
+                match value {
+                    Value::Str(string) => {
+                        let end = string.len();
+                        (string, 0..end)
+                    }
+                    _ => return exceptional_str_val_to_eval(name, value),
+                }
+            }
+            Evaluated::StrSlice { source, range } => (source, range),
+            Evaluated::Value(value) => match value {
+                Value::Str(string) => {
+                    let end = string.len();
+                    (string, 0..end)
+                }
+                _ => return exceptional_str_val_to_eval(name, value),
+            },
+        };
+
+        let start = {
+            let value = start.try_into()?;
+            match value {
+                Value::I64(num) => num,
+                _ => return exceptional_int_val_to_eval(name, value),
+            }
+        } - 1;
+
+        let count = match count {
+            Some(eval) => {
+                let value = eval.try_into()?;
+                match value {
+                    Value::I64(num) => num,
+                    _ => return exceptional_int_val_to_eval(name, value),
+                }
+            }
+            None => source.len() as i64,
+        };
+
+        let end = if count < 0 {
+            return Err(EvaluateError::NegativeSubstrLenNotAllowed.into());
+        } else {
+            (range.start as i64 + start + count).clamp(0, source.len() as i64) as usize
+        };
+
+        let start = (start + range.start as i64).clamp(0, source.len() as i64) as usize;
+
+        Ok(Evaluated::StrSlice {
+            source,
+            range: start..end,
+        })
+    }
+
+    pub fn trim(
+        self,
+        name: String,
+        filter_chars: Option<Evaluated<'_>>,
+        trim_where_field: &'a Option<TrimWhereField>,
```

Comment:
```suggestion
        trim_where_field: &Option<TrimWhereField>,
```

It has nothing to do with the lifetime of `TrimWhereField`.